src/pudl/analysis/allocate_net_gen.py

```diff
@@ -324,183 +509,437 @@ def stack_generators(
     return gens_stack
 
 
-def associate_generator_tables(gf, gen, gens):
-    """Associate the three tables needed to assign net gen to generators.
+def associate_generator_tables(
+    gens: pd.DataFrame, gf: pd.DataFrame, gen: pd.DataFrame, bf_by_gens: pd.DataFrame
+) -> pd.DataFrame:
+    """Associate the three tables needed to assign net gen and fuel to generators.
+
+    The ``generation_fuel_eia923`` table's data is reported at the
+    :py:const:`IDX_PM_ESC` granularity. Each generator has one ``prime_mover_code``, but
+    potentially several ``energy_source_code``s (reported in as primary source,
+    secondary source, etc.). Much of the data in ``generation_fuel_eia923`` not
+    associated with primary energy sources, so before merging the data with the
+    generators, we employ :func:``stack_generators``.
+
+    After the generator's energy source codes have been stacked, we group the other
+    data tables and merge them with the generators. We add suffixes to the data columns
+    to identify the source table.
+
+    This function also removes inactive generators so that we don't associate any net
+    generation or fuel to those generators. See :func:`remove_inactive_generators` for
+    more details.
+
+    There are some records in the data tables that have primary key values that do no
+    match up with any of those values in the generators table. We employ
+    :func:`_allocate_unassociated_records` to make sure those records are associated.
 
     Args:
-        gf (pandas.DataFrame): generator_fuel_eia923 table with columns:
-            ``IDX_PM_ESC`` and `net_generation_mwh` and `fuel_consumed_mmbtu`.
-        gen (pandas.DataFrame): generation_eia923 table with columns:
-            ``IDX_GENS`` and `net_generation_mwh`.
-        gens (pandas.DataFrame): generators_eia860 table with cols: ``IDX_GENS``
-            and all of the `energy_source_code` columns
-
-    TODO: Convert these groupby/merges into transforms.
+        gens: ``generators_eia860`` table with cols: :py:const:`IDX_GENS` and all of
+            the ``energy_source_code`` columns and expanded to the frequency of
+            ``pudl_out``
+        gf: generator_fuel_eia923 table with columns: :py:const:`IDX_PM_ESC` and
+            ``net_generation_mwh`` and ``fuel_consumed_mmbtu``.
+        gen: ``generation_eia923`` table with columns: :py:const:`IDX_GENS` and
+            ``net_generation_mwh``.
+        bf_by_gens: ``boiler_fuel_eia923`` table aggregated at the generator-level
+
+    Returns:
+        table of generators with stacked fuel types and broadcasted net generation data
+        from the generation_eia923 and generation_fuel_eia923 tables. There are many
+        duplicate values in this output.
```

Comment:
Clarify that the duplicate values are the desired / expected output, and they'll be allocated / deduplicated elsewhere (function ref?). It sounds like you might be noting a known error here.

Reply:
Also should mention boiler fuel data here.